pkg/kubelet/kubelet_pods.go

```diff
@@ -1595,6 +1596,30 @@ func (kl *Kubelet) convertStatusToAPIStatus(pod *v1.Pod, podStatus *kubecontaine
 		len(pod.Spec.InitContainers) > 0,
 		false,
 	)
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		//TODO(vinaykul): Move this code into a GenerateXXX function
+		specStatusDiffer := false
+		for _, c := range pod.Spec.Containers {
+			if _, cs, ok := podutil.GetContainerStatus(apiPodStatus.ContainerStatuses, c.Name); ok {
+				if cs.Resources != nil && diff.ObjectDiff(c.Resources, *cs.Resources) != "" {
+					specStatusDiffer = true
+					break
+				}
+			}
+		}
+		if !specStatusDiffer {
+			// Clear last resize state from checkpoint
+			if err := kl.statusManager.SetPodResizeStatus(pod.UID, ""); err != nil {
+				//TODO(vinaykul): Can we recover from this in some way? Investigate
```

Comment:
Will this leave a pod in an unrecoverable error state? That seems not great.

This mutation does not belong in this function, `convertStatusToAPIStatus` should only be transformative and not result in any mutation of the underlying pod status manager. This would break core workflow assumptions.

Reply:
Lift it one level up to `generateAPIPodStatus` https://github.com/kubernetes/kubernetes/blob/b90b2d963de46128e862995c912bc64b622eff11/pkg/kubelet/kubelet_pods.go#L1431

I'd expect the logic in the section below would probably be factored out into a helper.

Reply:
Sounds good. And about your earlier question, if the checkpoint write fails for some reason, it will temporarily leave the stored state with an inaccurate value. However, podStatus.Resize will have the right value (blank). Subsequent invocations of generateAPIPodStatus will retry updating checkpoint to the right value. GetPodResizeStatus is not invoked in any path where stale value gets picked up, so this error is inconsequential to the pod.
 
I think I added that TODO comment there at that time as a 'note to self' to ensure I come back and verify this was not an issue.

Reply:
The above change has been moved to generateAPIPodStatus a while back. I trust my above response answers the concern about unrecoverable state. Resolving this convo, please unresolve if anything is unclear.